internal/transport/transport_test.go

```diff
@@ -1663,81 +1663,279 @@ func (s) TestReadGivesSameErrorAfterAnyErrorOccurs(t *testing.T) {
+					success.Send(errors.New("the client received a frame other than Settings or Headers"))
+				}
 			}
-			switch frame := frame.(type) {
-			case *http2.SettingsFrame:
-				// Do nothing. A settings frame is expected from server preface.
-			case *http2.RSTStreamFrame:
-				if frame.Header().StreamID != 1 || http2.ErrCode(frame.ErrCode) != http2.ErrCodeProtocol {
-					// Client only created a single stream, so RST Stream should be for that single stream.
-					t.Errorf("RST stream received with streamID: %d and code %v, want streamID: 1 and code: http.ErrCodeFlowControl", frame.Header().StreamID, http2.ErrCode(frame.ErrCode))
+		}()
+
+		var buf bytes.Buffer
+		henc := hpack.NewEncoder(&buf)
+
+		// Needs to build headers deterministically to conform to gRPC over
+		// HTTP/2 spec.
+		for _, header := range test.headers {
+			for _, value := range header.values {
+				if err := henc.WriteField(hpack.HeaderField{Name: header.name, Value: value}); err != nil {
+					t.Fatalf("Error while encoding header: %v", err)
 				}
-				// Records that client successfully received RST Stream frame.
-				success.Send(nil)
-				return
-			default:
-				// The server should send nothing but a single RST Stream frame.
-				success.Send(errors.New("The client received a frame other than RST Stream"))
 			}
 		}
-	}()
-
-	// Done with HTTP/2 setup - now create a stream with a bad method header.
-	var buf bytes.Buffer
-	henc := hpack.NewEncoder(&buf)
-	// Method is required to be POST in a gRPC call.
-	if err := henc.WriteField(hpack.HeaderField{Name: ":method", Value: "PUT"}); err != nil {
-		t.Fatalf("Error while encoding header: %v", err)
-	}
-	// Have the rest of the headers be ok and within the gRPC over HTTP/2 spec.
-	if err := henc.WriteField(hpack.HeaderField{Name: ":path", Value: "foo"}); err != nil {
-		t.Fatalf("Error while encoding header: %v", err)
-	}
-	if err := henc.WriteField(hpack.HeaderField{Name: ":authority", Value: "localhost"}); err != nil {
-		t.Fatalf("Error while encoding header: %v", err)
-	}
-	if err := henc.WriteField(hpack.HeaderField{Name: "content-type", Value: "application/grpc"}); err != nil {
-		t.Fatalf("Error while encoding header: %v", err)
-	}
 
-	if err := framer.WriteHeaders(http2.HeadersFrameParam{StreamID: 1, BlockFragment: buf.Bytes(), EndHeaders: true}); err != nil {
-		t.Fatalf("Error while writing headers: %v", err)
-	}
-	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
-	defer cancel()
-	if e, err := success.Receive(ctx); e != nil || err != nil {
-		t.Fatalf("Error in frame server should send: %v. Error receiving from channel: %v", e, err)
+		if err := framer.WriteHeaders(http2.HeadersFrameParam{StreamID: 1, BlockFragment: buf.Bytes(), EndHeaders: true}); err != nil {
+			t.Fatalf("Error while writing headers: %v", err)
+		}
+		ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+		defer cancel()
+		if e, err := success.Receive(ctx); e != nil || err != nil {
```

Comment:
Same suggestions as above.

Reply:
Changed "success" to "result". Added error checks the same as the other.